### DISCUSSION TITLE
DRT428M-2 rs485 MODBus

### DISCUSSION THREAD

**Discussion by rieders** [arendst/Tasmota#14746] [Show and tell]
Hi
I've been trying to integrate a DRT428M-2 into Tasmota for a few days.
I tried it with Tasmota sensors and Tasmota Smart Meters.
At Tasmota Smart Meters I have
#ifndef USE_SCRIPT
#define USE_SCRIPT
#endif
#ifndef USE_SML_M
#define USE_SML_M
#endif
#ifdef USE_RULES
#undef USE_RULES
#endif
integrated.
And the script
>DH
>B
->sensor53 r
>M 1
+1,3,M,1,9600,DRT,1,5,01030003
Used to see if the smart meter is responding.
Unfortunately I get no reaction.
Even if I set mE1 I get no answer.
Do I have to activate 8E1 somewhere else?

I get a nice list of all the values via the software supplied with the meter.

![2](https://user-images.githubusercontent.com/5175143/152679595-7079071b-d342-4b6a-a96e-1aee0d64e84a.PNG)
![3](https://user-images.githubusercontent.com/5175143/152679600-89194dc9-aafd-4bc9-9399-ecafbc0e43e7.PNG)
![Unbenannt](https://user-images.githubusercontent.com/5175143/152679604-57316fd1-3abe-4943-a407-2237852acb13.PNG)

I would be happy if someone could help me.

Regards
Andre

**Comment by barbudor** [comment]
Hi
First could you please confirm which RS485 adapter you are using with Tasmota. Tasmota doesn't support adapters that need a DIRection pin to be provided. It must support automatic direction. Exemple of such supported adapter : https://www.aliexpress.com/item/4000435224794.html

Which board re you using ?
On NodeMCU and Mini-D1, if you want to use GPIO 3 as RX, USB must be disconnected as GPIO3 (RX) and GPIO1 (TX) are connected to the USB/Serial adapter.
Alternatively you can use GPIO13 (RX) and GPIO15 (TX). Tasmota should be able to use those pins with the hardware UART.

**Comment by rieders** [comment]
Hi

Many thanks for the quick response.
I have this converter in the link.
I tried Tasmota on Wemos D1 mini and MCU LoLin.
Yes, it can be sent. I just don't receive anything.

**Comment by timaseth** [comment]
Please see if your TX and RX connections are straight and _not_ cross-over (which is the usual case)
So, what you want is:
TX >> TX
RX >> RX

This is what I have and it works very well:
```
>D 
>B
=>sensor53 r
>M 1
+1,3,m,0,9600,EM1,1,1,[followed by the modes addresses you want]
```

**Comment by rieders** [comment]
Hi
Many thanks for the help.
Unfortunately I didn't get any results.
No result is displayed in the console.
I've tried many options now without success.
>D
>B
=>sensor53 r
>M 1
+1.0,M,1.9600,DRT,2.5.0103000e
1.01040UUuu@i0:1,Voltage L1,V,Voltage_L1,0

#
Unfortunately, your script does not produce any results either.
Somehow it must be possible!

Regards
Andre

**Reply by timaseth** [reply]
> Hi Many thanks for the help. Unfortunately I didn't get any results. No result is displayed in the console. I've tried many options now without success.
> 
> > D
> > B
> > =>sensor53 r
> > M 1
> > +1.0,M,1.9600,DRT,2.5.0103000e
> > 1.01040UUuu@i0:1,Voltage L1,V,Voltage_L1,0
> 
> Unfortunately, your script does not produce any results either. Somehow it must be possible!
> 
> Regards Andre

Can you share your exact wiring diagram and the verbatim script in use? Perhaps a screen grab of the script, if you please.

**Comment by barbudor** [comment]
Do you have an oscilloscope to check that you are actually receiving data ?

You may also disable the script and set the GPIOs 1 / 3  to  Serial TX et Serial RX like below

![image](https://user-images.githubusercontent.com/2996042/152699303-11b484a9-0797-4ce4-8a69-0fafb3ee0e53.png)

Commands : https://tasmota.github.io/docs/Commands/#serial-bridge
`Baudrate 9600`
`SerialConfig 8E1`  
Then use `SerialSend5 XXXXXX` with hexdata (like `SerialSend5 0155AA`) to request data from the meter
Even if the meter send data on it's own, you need to send at least 1 dummy data to enable reception

**Comment by rieders** [comment]
Hi

Many thanks for the support.
I measured the outputs with an oscilloscope. They have a signal.
What I did is try to find out why the signal is not arriving with software (Serial Port Monitor).
Strangely it says that address 48 is answered. But I have set address 1.

I'll send to address 48 tomorrow.
Let's see if it works then.

![4](https://user-images.githubusercontent.com/5175143/152830085-6e510832-b5a0-41cf-a0e8-5a2decb604ef.PNG)


Regards
Andre

**Comment by rieders** [comment]
Hi
Today I connected a 2 RS485 converter to read the sent data to the Modbus as a test.
Unfortunately I can't do much with the data.
Maybe someone knows what the script can be called so that I can receive the data from the electricity meter.

![1](https://user-images.githubusercontent.com/5175143/153040068-e1fe0dd0-849a-4cbd-ae12-87f7c6f766b0.PNG)
![2](https://user-images.githubusercontent.com/5175143/153040099-47136ea6-c20c-4e27-9c00-be3e6cbbc8cc.PNG)
![3](https://user-images.githubusercontent.com/5175143/153040118-1ad77415-7b1c-4a6c-98ad-a4eaf1c86ada.PNG)

>DH
>B
=>sensor53 r
>M 1

+1,3,M,1,9600,DRT,1,2,0103000e
1,010304UUuu@i0:1,Spannung L1,V,Voltage_L1,0


#


Greetings Andre

**Reply by gemu2015** [reply]
you should go to console and type sensor53 d1
this activates dump mode. now you should see the cmds going to the meter and the response from the meter.
when this is working deactivate dump mode again with sensor53 d0
then from the register list of your meter your registers are float numbers
so you should use
1,010304ffffffff@i0:1,Spannung L1,V,Voltage_L1,0

and if you get the right value add more registers to the descriptor.
in the SML docs there are a couple of examples.

**Comment by rieders** [comment]
Hi

The problem I have is that the master (electricity meter) is not responding.
I noticed that the Serial Port monitor was only receiving data from Modbus RTU.
This differs from ASCII in that it requires 3.5 characters of waiting time, while ASCII only needs one character.
How can I change this in the script?
Or does something have to be adjusted in Tasmota.

Regards
Andre

**Comment by turutzi** [comment]
Stuck whit the same problem. -.-
I have tried a few things in the code but my programming skills arent that good.
Maybe is "Modbus ASCII" the awnser to the thing.

**Comment by ahermann86** [comment]
Hello I found the problem! 
The counter wants **2 stop bits**. 
I use it with FHEM via the "Serial to TCP Bridge / #define USE_TCP_BRIDGE" 
I'm just failing at the moment to set this in the Tasmota/ESP... 
I solved it for myself with a Quick and Dirty method ;)

File xdrv_41_tcp_bridge.ino with yellow marked modifications:
![grafik](https://user-images.githubusercontent.com/48262831/162574068-50233b2e-24f6-4475-ae9b-5766dad88a32.png)

This is how it looks in FHEM:
![grafik](https://user-images.githubusercontent.com/48262831/162574177-cd7c441e-00b1-46b9-884c-49162eb4a4b8.png)

**Reply by barbudor** [reply]
Wouldn't [`TcpConfig 8N2`](https://tasmota.github.io/docs/Commands/#tcpconfig) have done the trick ?

**Comment by ahermann86** [comment]
I didn't get the idea because this  [documentation ](https://tasmota.github.io/docs/Serial-to-TCP-Bridge/) describes it that way: "TCPBaudRate <x>: sets the baudrate for serial (only 8N1 mode)"

But you're right, it works with that too.

Thank you 👍

**Reply by barbudor** [reply]
Indeed this doc page need to be amended to add the command TcpConfig

**Comment by gemu2015** [comment]
if that is the reason you may use this in SML driver to set MODBUS to 8E2:

+1,3,ME2,1,9600,DRT,1,2,0103000e

**Comment by naseweiss01** [comment]
Despite all speculation, my DRT 428M-2 is working with 8E1 !!!
>D
>B
->sensor53 r
>M 1

+1,3,ME1,1,9600,DRT,1,3,0103000E,01030010,01030012,01030016,0103001C,01030100
1,010304ffffffff@i0:1,Voltage L1,V,Voltage_L1,1
1,010304ffffffff@i1:1,Voltage L2,V,Voltage_L2,1
1,010304ffffffff@i2:1,Voltage L3,V,Voltage_L3,1
1,010304ffffffff@i3:1,Current L1,A,Current_L1,1
1,010304ffffffff@i4:0.001,Power tot,W,Power_tot,1
1,010304ffffffff@i5:1,Energie_tot,kWh,Energie_tot,1
#

![image](https://user-images.githubusercontent.com/29761535/205434574-f953a715-9716-4aee-bfa8-19438b77284a.png)
